### PR TITLE
Simplify InterceptedMethods. Fix #439

### LINKED ISSUE #439: x == y fails with a NullPointerException when x is null
`InterceptedMethods` now rewrite `x == y` into `x.equals(y)` but this means that if x is null, there will be a NullPointerException, surely that's not the intended behaviour?

### PR DESCRIPTION
**Pull request by DarkDimius**
It's now a single pattern-match, instead of series of nested ifs.
review by @smarter

### CODE REVIEW

File: src/dotty/tools/dotc/transform/InterceptedMethods.scala [hunk 1]
```diff
@@ -99,47 +98,42 @@ class InterceptedMethods extends MiniPhaseTransform { thisTransform =>
...
-          } else if (Any_comparisons contains tree.fun.symbol.asTerm) {
-            if (tree.fun.symbol eq defn.Any_==) {
-              qual.selectWithSig(defn.Any_equals).appliedToArgs(tree.args)
-            } else if (tree.fun.symbol eq defn.Any_!=) {
-              qual.selectWithSig(defn.Any_equals).appliedToArgs(tree.args).select(defn.Boolean_!)
-            } else unknown
-          } /* else if (isPrimitiveValueClass(qual.tpe.typeSymbol)) {
+    lazy val Select(qual, _) = tree.fun
+    val Any_## = this.Any_##
+    val Any_!= = this.Any_!=
```
Comment (smarter) [inline src/dotty/tools/dotc/transform/InterceptedMethods.scala]: Why not:
```scala
val Any_## = defn.Any_##
val Any_!= = defn.Any_!=
```

File: src/dotty/tools/dotc/transform/InterceptedMethods.scala [hunk 2]
```diff
@@ -99,47 +98,42 @@ class InterceptedMethods extends MiniPhaseTransform { thisTransform =>
 
   override def transformApply(tree: Apply)(implicit ctx: Context, info: TransformerInfo): Tree = {
     def unknown = {
-      assert(false, s"The symbol '${tree.fun.symbol.showLocated}' was intercepted but didn't match any cases, " +
+      assert(false, s"The symbol '${ tree.fun.symbol.showLocated }' was intercepted but didn't match any cases, " +
```
Comment (smarter) [inline src/dotty/tools/dotc/transform/InterceptedMethods.scala]: Why the extra spaces here?

File: src/dotty/tools/dotc/transform/InterceptedMethods.scala [hunk 3]
```diff
@@ -99,47 +98,42 @@ class InterceptedMethods extends MiniPhaseTransform { thisTransform =>
...
-            unknown
-          }
-        case _ =>
-          unknown
-      }
-      ctx.log(s"$phaseName rewrote $tree to $rewrite")
-      rewrite
+         */
+      case t if primitiveGetClassMethods.contains(t) =>
+                  // if we got here then we're trying to send a primitive getClass method to either
```
Comment (smarter) [inline src/dotty/tools/dotc/transform/InterceptedMethods.scala]: Unnecessary extra indentation

File: src/dotty/tools/dotc/transform/InterceptedMethods.scala (hunk 2 above)
Comment (DarkDimius) [inline src/dotty/tools/dotc/transform/InterceptedMethods.scala]: IntelliJ made this silently.

**Comment by smarter** [comment]
The documentation for the class says:
> * - `x == y` for == in class Any becomes `x equals y` with equals in class Object.
But this is no longer the case, this comment needs to be updated.

**Comment by smarter** [comment]
I wonder if `poundPoundValue` could be rewritten with `applyOverloaded`.

File: src/dotty/tools/dotc/transform/InterceptedMethods.scala (hunk 2 above)
Comment (smarter) [inline src/dotty/tools/dotc/transform/InterceptedMethods.scala]: `unknown` is no longer used, it should either be used or removed.

**Comment by DarkDimius** [comment]
It could. But I would better leave it as is. We do not need full overload resolution(which could be resource-intensive)

File: src/dotty/tools/dotc/transform/InterceptedMethods.scala [hunk 4]
```diff
@@ -28,42 +28,32 @@ import dotty.tools.dotc.core.Denotations.SingleDenotation
...
-// @DarkDimius The getClass scheme changed. We no longer can have
-// two different methods in Any and Object. The tests pass but I
-// am not sure Intercepted methods treats getClass right now.
-// Please check and delete comment when done.
 /** Replace member references as follows:
   *
-  * - `x == y` for == in class Any becomes `x equals y` with equals in class Object.
-  * - `x != y` for != in class Any becomes `!(x equals y)` with equals in class Object.
-  * - `x.##` for ## in other classes becomes calls to ScalaRunTime.hash,
+  * - `x != y` for != in class Any becomes `!(x != y)` with != in class Any.
```
Comment (smarter) [inline src/dotty/tools/dotc/transform/InterceptedMethods.scala]: Typo, should be !(x == y) with == in ...

Comment (DarkDimius) [inline src/dotty/tools/dotc/transform/InterceptedMethods.scala]: yes. Thanks for noticing. 
I'd better go and have some rest.

**Comment by smarter** [comment]
LGTM

**Pull request merged**